src/main/scala/io/iohk/ethereum/consensus/ConsensusImpl.scala

```diff
@@ -1,81 +1,131 @@
+    *   - [[BlockImportFailed]]  - block failed to execute (when importing to top or reorganising the chain)
     */
-  def importBlock(
+  override def evaluateBranchBlock(
       block: Block
   )(implicit blockExecutionScheduler: Scheduler, blockchainConfig: BlockchainConfig): Task[BlockImportResult] =
     blockchainReader.getBestBlock() match {
       case Some(bestBlock) =>
         if (isBlockADuplicate(block.header, bestBlock.header.number)) {
-          Task(log.debug(s"Ignoring duplicate block: (${block.idTag})"))
-            .map(_ => DuplicateBlock)
+          log.debug("Ignoring duplicated block: {}", block.idTag)
+          Task.now(DuplicateBlock)
         } else {
-          val hash = bestBlock.header.hash
-          blockchain.getChainWeightByHash(hash) match {
+          blockchain.getChainWeightByHash(bestBlock.header.hash) match {
             case Some(weight) =>
-              val importResult = if (isPossibleNewBestBlock(block.header, bestBlock.header)) {
-                importToTop(block, bestBlock, weight)
-              } else {
-                reorganise(block, bestBlock, weight)
+              doBlockPreValidation(block).flatMap {
+                case Left(error) => Task.now(BlockImportFailed(error.reason.toString))
+                case Right(_)    => handleBlockImport(block, bestBlock, weight)
               }
-              importResult.foreach(measureBlockMetrics)
-              importResult
-            case None =>
-              log.error(
-                "Getting total difficulty for current best block with hash: {} failed",
-                bestBlock.header.hashAsHexString
-              )
-              Task.now(
-                BlockImportFailed(
-                  s"Couldn't get total difficulty for current best block with hash: ${bestBlock.header.hashAsHexString}"
-                )
-              )
+            case None => returnNoTotalDifficulty(bestBlock)
           }
         }
-      case None =>
-        log.error("Getting current best block failed")
-        Task.now(BlockImportFailed("Couldn't find the current best block"))
+      case None => returnNoBestBlock()
     }
 
+  private def handleBlockImport(block: Block, bestBlock: Block, weight: ChainWeight)(implicit
+      blockExecutionScheduler: Scheduler,
+      blockchainConfig: BlockchainConfig
+  ): Task[BlockImportResult] = {
+    val importResult = if (isPossibleNewBestBlock(block.header, bestBlock.header)) {
+      importToTop(block, bestBlock, weight)
+    } else {
+      reorganiseOrEnqueue(block, bestBlock, weight)
+    }
+    importResult.foreach(measureBlockMetrics)
+    importResult
+  }
+
+  private def doBlockPreValidation(block: Block)(implicit
+      blockchainConfig: BlockchainConfig
+  ): Task[Either[ValidationBeforeExecError, BlockExecutionSuccess]] =
+    Task
+      .evalOnce {
+        val validationResult = blockValidation.validateBlockBeforeExecution(block)
+        validationResult.left.foreach { error =>
```

Comment:
While at it this can be improved using `tap` from cats-effect.

Reply:
Thanks! Done